homeassistant/helpers/config_entry_oauth2_flow.py

```diff
@@ -0,0 +1,420 @@
+        self.flow_impl: AbstractOAuth2Implementation = None  # type: ignore
+
+    @property
+    @abstractmethod
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+
+    @property
+    def extra_authorize_data(self) -> dict:
+        """Extra data that needs to be appended to the authorize url."""
+        return {}
+
+    async def async_step_pick_implementation(self, user_input: dict = None) -> dict:
+        """Handle a flow start."""
+        assert self.hass
+        implementations = await async_get_implementations(self.hass, self.DOMAIN)
+
+        if user_input is not None:
+            self.flow_impl = implementations[user_input["implementation"]]
+            return await self.async_step_auth()
+
+        if not implementations:
+            return self.async_abort(reason="missing_configuration")
+
+        if len(implementations) == 1:
+            # Pick first implementation as we have only one.
+            self.flow_impl = list(implementations.values())[0]
+            return await self.async_step_auth()
+
+        return self.async_show_form(
+            step_id="pick_implementation",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        "implementation", default=list(implementations.keys())[0]
+                    ): vol.In({key: impl.name for key, impl in implementations.items()})
+                }
+            ),
+        )
+
+    async def async_step_auth(self, user_input: dict = None) -> dict:
+        """Create an entry for auth."""
+        # Flow has been triggered by external data
+        if user_input:
+            self.external_data = user_input
+            return self.async_external_step_done(next_step_id="creation")
+
+        try:
+            with async_timeout.timeout(10):
+                url = await self.flow_impl.async_generate_authorize_url(self.flow_id)
+        except asyncio.TimeoutError:
+            return self.async_abort(reason="authorize_url_timeout")
+
+        url = str(URL(url).update_query(self.extra_authorize_data))
+
+        return self.async_external_step(step_id="auth", url=url)
+
+    async def async_step_creation(self, user_input: dict = None) -> dict:
+        """Create config entry from external data."""
+        token = await self.flow_impl.async_resolve_external_data(self.external_data)
+        token["expires_at"] = time.time() + token["expires_in"]
+
+        self.logger.info("Successfully authenticated")
+
+        return await self.async_oauth_create_entry(
+            {"auth_implementation": self.flow_impl.domain, "token": token}
+        )
+
+    async def async_oauth_create_entry(self, data: dict) -> dict:
+        """Create an entry for the flow.
+
+        Ok to override if you want to fetch extra info or even add another step.
+        """
+        return self.async_create_entry(title=self.flow_impl.name, data=data)
+
+    async_step_user = async_step_pick_implementation
```

Comment:
Is user the correct step for this? Perhaps auth? There are apis where user action is needed before the authorization, like setting scopes or in withings case, selecting which profile the callback will associate with the data.

Reply:
`user` is the entrypoint for config flows if launched by a user. You can override the `async_step_user` to do what you want, just make sure you navigate to `pick_implementation` when done. For an example, see the gist that I put in the description: https://gist.github.com/balloob/e2abc99f0036eed4f308f19d36be9ac9

I also think that profile picking can be done after the linking is done, that would make more sense as we can see which profiles are available.